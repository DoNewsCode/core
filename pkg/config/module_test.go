// +build core

package config

import (
	"github.com/DoNewsCode/std/pkg/container"
	"github.com/DoNewsCode/std/pkg/contract"
	"github.com/spf13/cobra"
	"github.com/stretchr/testify/assert"
	"io/ioutil"
	"os"
	"testing"
)

func setup() *cobra.Command {
	os.Remove("./testdata/module_test.yaml")
	os.Remove("./testdata/module_test.json")
	var cont container.Container
	var config, _ = NewConfig()
	var mod = Module{Container: &cont}
	cont.AddModule(MockModule(func() []contract.ExportedConfig {
		return []contract.ExportedConfig{
			{
				"foo",
				map[string]interface{}{
					"foo": "bar",
				},
				"A mock config",
			},
		}
	}))
	cont.AddModule(Module{Container: &cont, Conf: config})
	rootCmd := &cobra.Command{
		Use: "root",
	}
	mod.ProvideCommand(rootCmd)
	return rootCmd
}

type MockModule func() []contract.ExportedConfig

func (m MockModule) ProvideConfig() []contract.ExportedConfig {
	return m()
}

func TestModule_ProvideCommand(t *testing.T) {
	rootCmd := setup()
	cases := []struct {
		name string
		args []string
	}{
		{
			"new yaml",
			[]string{"exportConfig", "--outputFile", "testdata/module_test.yaml"},
		},
		{
			"old yaml",
			[]string{"exportConfig", "--outputFile", "testdata/module_test.yaml"},
		},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			rootCmd.SetArgs(c.args)
			rootCmd.Execute()
			testTarget, _ := ioutil.ReadFile("testdata/module_test.yaml")
			expected, _ := ioutil.ReadFile("testdata/module_test_expected.yaml")
			assert.Equal(t, expected, testTarget)
		})
	}
	cases = []struct {
		name string
		args []string
	}{
		{
			"new json",
			[]string{"exportConfig", "--outputFile", "testdata/module_test.json", "--style", "json"},
		},
		{
			"old json",
			[]string{"exportConfig", "--outputFile", "testdata/module_test.json", "--style", "json"},
		},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			rootCmd.SetArgs(c.args)
			rootCmd.Execute()
			testTarget, _ := ioutil.ReadFile("testdata/module_test.json")
			expected, _ := ioutil.ReadFile("testdata/module_test_expected.json")
			assert.Equal(t, expected, testTarget)
		})
	}
}
