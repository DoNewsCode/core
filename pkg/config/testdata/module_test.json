{
  "env": "local",
  "foo": "bar",
  "grpc": {
    "addr": ":9090"
  },
  "http": {
    "addr": ":8080"
  },
  "name": "app",
  "version": "0.1.0"
}
