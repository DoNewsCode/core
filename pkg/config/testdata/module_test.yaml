# A mock config
foo: bar

# The name of the application
name: app

# The version of the application
version: 0.1.0

# The environment of the application, one of production, development, staging, testing or local
env: local

# The http address
http:
  addr: :8080

# The gRPC address
grpc:
  addr: :9090

