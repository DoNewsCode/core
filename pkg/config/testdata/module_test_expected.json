{
  "env": "local",
  "foo": "bar",
  "grpc": {
    "addr": ":9090"
  },
  "http": {
    "addr": ":8080"
  },
  "name": "skeleton",
  "version": "0.1.0"
}
